magicgui/type_map.py

```diff
@@ -369,6 +380,58 @@ def _deco(type_):
     return _deco if type_ is None else _deco(type_)
 
 
+@contextmanager
+def type_registered(
+    type_: _T,
+    *,
+    widget_type: WidgetRef | None = None,
+    return_callback: ReturnCallback | None = None,
+    **options,
+) -> Iterator[None]:
+    """Context manager that temporarily registers a widget type for a given `type_`.
+
+    When the context is exited, the previous widget type associations for `type_` is
+    restored.
+
+    Parameters
+    ----------
+    type_ : _T
+        The type for which a widget class or return callback will be provided.
+    widget_type : Optional[WidgetRef]
+        A widget class from the current backend that should be used whenever ``type_``
+        is used as the type annotation for an argument in a decorated function,
+        by default None
+    return_callback: Optional[callable]
+        If provided, whenever ``type_`` is declared as the return type of a decorated
+        function, ``return_callback(widget, value, return_type)`` will be called
+        whenever the decorated function is called... where ``widget`` is the Widget
+        instance, and ``value`` is the return value of the decorated function.
+    **options
+        key value pairs where the keys are valid `WidgetOptions`
+    """
+    tw = TypeWrapper(type_)
+    tw.resolve()
+    _type_ = tw.outer_type_
+    null = object()
+
+    prev_return_callback = _RETURN_CALLBACKS.get(_type_, null)
+    prev_type_def = _TYPE_DEFS.get(_type_, null)
+    try:
+        register_type(
+            type_, widget_type=widget_type, return_callback=return_callback, **options
+        )
```

Comment:
having this under `try` clause is to keep the consistency of the data structure. If yes, then maybe this should be done in `register_type` itself?

Reply:
i agree, bit odd in the try block there.  gonna just move it above.